packages/zoe/src/contractSupport/ratio.js

```diff
@@ -0,0 +1,172 @@
+  );
+  for (const name of propertyNames) {
+    assert(
+      ratioPropertyNames.includes(name),
+      X`Parameter must be a Ratio record, but ${ratio} has ${name}`,
+    );
+  }
+  Nat(ratio.numerator);
+  Nat(ratio.denominator);
+}
+
+export function makeRatio(
+  numerator,
+  numeratorBrand,
+  denominator = PERCENT,
+  denominatorBrand = numeratorBrand,
+) {
+  assert(denominator > 0, X`No infinite ratios!`);
+
+  return harden({
+    numerator: Nat(numerator),
+    denominator: Nat(denominator),
+    numeratorBrand,
+    denominatorBrand,
+  });
+}
+
+export function makeRatioFromAmounts(
+  numeratorAmount,
+  denominatorAmount,
+  base = BASIS_POINTS,
+) {
+  assert(denominatorAmount.value > 0, X`No infinite ratios!`);
+
+  const numerator = floorDivide(
+    multiply(Nat(base), numeratorAmount.value),
+    denominatorAmount.value,
+  );
+  return harden({
+    numerator,
+    denominator: Nat(base),
+    numeratorBrand: numeratorAmount.brand,
+    denominatorBrand: denominatorAmount.brand,
+  });
+}
+
+export function multiplyBy(amount, ratio) {
+  assertIsRatio(ratio);
+  assert(
+    amount.brand === ratio.denominatorBrand,
+    X`amount's brand ${amount.brand} must match ratio's denominator ${ratio.denominatorBrand}`,
+  );
+
+  return harden({
+    value: floorDivide(
+      multiply(amount.value, ratio.numerator),
+      ratio.denominator,
+    ),
+    brand: ratio.numeratorBrand,
+  });
+}
+
+export function divideBy(amount, ratio) {
+  assertIsRatio(ratio);
+  assert(
+    amount.brand === ratio.numeratorBrand,
+    X`amount's brand ${amount.brand} must match ratio's numerator ${ratio.numeratorBrand}`,
+  );
+  return harden({
+    value: floorDivide(
+      multiply(amount.value, ratio.denominator),
+      ratio.numerator,
+    ),
+    brand: ratio.denominatorBrand,
+  });
+}
+
+export function invertRatio(ratio) {
+  assertIsRatio(ratio);
+
+  return makeRatio(
+    ratio.denominator,
+    ratio.denominatorBrand,
+    ratio.numerator,
+    ratio.numeratorBrand,
+  );
+}
+
+export function multiplyRatios(ratioA, ratioB) {
+  assertIsRatio(ratioA);
+  assertIsRatio(ratioB);
+
+  if (ratioA.numeratorBrand === ratioB.denominatorBrand) {
+    return makeRatio(
+      multiply(ratioA.numerator, ratioB.numerator),
+      ratioB.numeratorBrand,
+      multiply(ratioA.denominator, ratioB.denominator),
+      ratioA.denominatorBrand,
+    );
+  } else if (ratioA.denominatorBrand === ratioB.numeratorBrand) {
+    return makeRatio(
+      multiply(ratioA.numerator, ratioB.numerator),
+      ratioA.numeratorBrand,
+      multiply(ratioA.denominator, ratioB.denominator),
+      ratioB.denominatorBrand,
+    );
+  }
+  assert.fail(X`Ratios must have a common unit`);
+}
+
+// ///// PERCENT ///////////////////
+
+// If ratio is between 0 and 1, subtract from 1.
+export function complementPercent(ratio) {
+  assertIsRatio(ratio);
+  assert(
+    ratio.numerator <= ratio.denominator,
+    X`Ratio must be less than 1 to take its complement`,
```

Comment:
Test is less that or equal but details message says less than

Reply:
fixed